test/topology_experimental_raft/test_coordinator_aborting.py

```diff
@@ -0,0 +1,183 @@
+import pytest
+import asyncio
+import logging
+import re
+import time
+from test.pylib.internal_types import ServerInfo
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.scylla_cluster import Cluster, ReplaceConfig
+from test.pylib.util import wait_for_cql_and_get_hosts, read_barrier
+from cassandra.cluster import SimpleStatement, ConsistencyLevel
+from test.topology.util import check_token_ring_and_group0_consistency
+
+
+logger = logging.getLogger(__name__)
+
+
+async def restart_coordinator_after_log_message_on_target_node(manager: ManagerClient, log_pattern: str, coordinator: ServerInfo, 
+                              verification_node: ServerInfo, target_node: ServerInfo):
+    """ Restart coordinator node after log_pattern found in logs on node_under_ops
+    
+    Wait for pattern is found in logs on target_node, then stop coordinator, 
+    validate with verification_node, that coordinator was stopped (trigger election
+    of new coordinator), start coordinator node
+    """
+    log = await manager.server_open_log(target_node.server_id)
+    await log.wait_for(log_pattern, timeout=30)
+    await manager.server_stop(coordinator.server_id)
+    await manager.server_not_sees_other_server(verification_node.ip_addr, coordinator.ip_addr)
+    await manager.server_start(coordinator.server_id)
+
+
+@pytest.mark.asyncio
+async def test_kill_coordinator_during_op(manager: ManagerClient) -> None:
+    """ Kill coordinator while topology operation is running for cluster: decommission,
+    bootstrap, removenode, replace.
+
+    1. Start operation on target node
+    2. Wait for appropriate message is in log on target node
+    3. Stop coordinator node, after log pattern found in log on target node
+    4. Wait for new coordinator will be elected
+    5. Start old coordinator node
+
+    Topology operation is expected to fail and cluster is rolled back
+    to previous state.    
+    """
+    # Decrease the failure detector threshold so we don't have to wait for too long.
+    config = {
+        'failure_detector_timeout_in_ms': 2000
+    }
+    nodes = [await manager.server_add(config=config) for _ in range(4)]
+    cql = manager.get_cql()
+
+    coordinators_ids = await get_coordinator_host_ids(manager)
+    assert len(coordinators_ids) == 1, "At least 1 coordinator id should be found"
+    # kill coordinator during decommission
+    coordinator_host = await get_coordinator_host(coordinators_ids[0], manager)
+    other_nodes = [srv for srv in nodes if srv.server_id != coordinator_host.server_id]
+
+    await asyncio.gather(manager.decommission_node(server_id=other_nodes[-1].server_id,
+                                                                     expected_error="Decommission failed. See earlier errors"),
+                         restart_coordinator_after_log_message_on_target_node(manager, 
```

Comment:
Did anything change here? This is still susceptible to races, and won't work every time.

You're enforcing an order between these two events:
E1: target node starts decommission (by printing "request decommission for ...")
E2: coordinator is restarted

We can write this in short: E1 < E2.

But there is nothing that guarantees when decommission will end or fail.
Let's say that
E3: decommission ends

The test does not do anything to ensure that E2 < E3. It could happen that decommission ends before coordinator is restarted.

And even if you had:
E1 < E2 < E3

it would still not be enough: coordinator could get restarted at a point where decommission no longer needs coordinator to succeed, i.e. coordinator has finished its work, and sends message to target node telling it that decommission has succeeded, and then the test restarts it. In that case decommission still won't fail.

You need to somehow guarantee that coordinator is restarted **after decommission starts**, but **before target node stops needing the coordinator node**.

One way to ensure that, which comes to my mind, is to have an injection in the target node's code that will pause the node when coordinator is waiting for it. For example, when coordinator sends stream request to node and waits for it to finish streaming. Inside streaming code you could have an injection that pauses the node and waits for message from test. The injection would print a log message then pause the node. The test would wait for this message, then kill the coordinator, then send message to target node to unpause it. This would enforce an order of the following events:
E1: target node starts decommission (by printing "request decommission for ...")
E2: coordinator is restarted
E3: target node sends streaming finished message to coordinator

E1 < E2 < E3

Then we could expect decommission to fail.